CardPayments/src/main/java/com/paypal/android/cardpayments/CardAuthLauncher.kt

```diff
@@ -0,0 +1,153 @@
+        options: BrowserSwitchOptions
+    ): PayPalSDKError? {
+        var error: PayPalSDKError? = null
+        try {
+            browserSwitchClient.start(activity, options)
+        } catch (e: BrowserSwitchException) {
+            error = CardError.browserSwitchError(e)
+        }
+        return error
+    }
+
+    fun deliverBrowserSwitchResult(activity: FragmentActivity) =
+        browserSwitchClient.deliverResult(activity)?.let { browserSwitchResult ->
+            val requestType =
+                browserSwitchResult.requestMetadata?.optString(METADATA_KEY_REQUEST_TYPE)
+            if (requestType == REQUEST_TYPE_VAULT) {
+                parseVaultResult(browserSwitchResult)
+            } else {
+                // Assume REQUEST_TYPE_APPROVE_ORDER
+                parseApproveOrderResult(browserSwitchResult)
+            }
+        }
+
+    private fun parseVaultResult(browserSwitchResult: BrowserSwitchResult): CardStatus? {
+        val setupTokenId =
+            browserSwitchResult.requestMetadata?.optString(METADATA_KEY_SETUP_TOKEN_ID)
+        return when (browserSwitchResult.status) {
+            BrowserSwitchStatus.SUCCESS -> parseVaultSuccessResult(browserSwitchResult)
+            BrowserSwitchStatus.CANCELED -> CardStatus.VaultCanceled(setupTokenId)
+            else -> null
+        }
+    }
+
+    private fun parseApproveOrderResult(browserSwitchResult: BrowserSwitchResult): CardStatus? {
+        val orderId = browserSwitchResult.requestMetadata?.optString(METADATA_KEY_ORDER_ID)
+        return if (orderId == null) {
+            CardStatus.ApproveOrderError(CardError.unknownError)
+        } else {
+            when (browserSwitchResult.status) {
+                BrowserSwitchStatus.SUCCESS ->
+                    parseApproveOrderSuccessResult(browserSwitchResult, orderId)
+
+                BrowserSwitchStatus.CANCELED -> CardStatus.ApproveOrderCanceled(orderId)
+                else -> null
+            }
+        }
+    }
+
+    private fun parseVaultSuccessResult(browserSwitchResult: BrowserSwitchResult): CardStatus {
+        val deepLinkUrl = browserSwitchResult.deepLinkUrl
+        val requestMetadata = browserSwitchResult.requestMetadata
+
+        return if (deepLinkUrl == null || requestMetadata == null) {
+            CardStatus.VaultError(CardError.unknownError)
+        } else {
+            // TODO: see if there's a way that we can require the merchant to make their
+            // return and cancel urls conform to a strict schema
+
+            // NOTE: this assumes that when the merchant created a setup token, they used a
+            // return_url with word "success" in it (or a cancel_url with the word "cancel" in it)
+            val setupTokenId =
+                browserSwitchResult.requestMetadata?.optString(METADATA_KEY_SETUP_TOKEN_ID)
+            val deepLinkUrlString = deepLinkUrl.toString()
+            val didSucceed = deepLinkUrlString.contains("success")
```

Comment:
This was the best way I could figure out if 3DS was successful. If we keep it this way, we will need to call out to merchants that their URL needs to have "success" in it.